include/envoy/ssl/handshaker.h

```diff
@@ -43,5 +46,81 @@ class Handshaker {
   virtual Network::PostIoAction doHandshake() PURE;
 };
 
+using HandshakerSharedPtr = std::shared_ptr<Handshaker>;
+using HandshakerFactoryCb =
+    std::function<HandshakerSharedPtr(bssl::UniquePtr<SSL>, int, HandshakeCallbacks*)>;
+
+class HandshakerFactoryContext {
+public:
+  virtual ~HandshakerFactoryContext() = default;
+
+  /**
+   * @return reference to the Api object
+   */
+  virtual Api::Api& api() PURE;
+
+  /**
+   * The list of supported protocols exposed via ALPN, from ContextConfig.
+   */
+  virtual absl::string_view alpnProtocols() const PURE;
+};
+
+struct HandshakerRequirements {
+  // Should be true if a handshaker requires certificates in the config.
+  bool require_certificates = true;
+
+  // Should be true if a handshaker requires that Envoy sets the session
+  // resumption callback described by `SSL_CTX_set_select_certificate_cb`.
+  bool should_set_select_certificate_cb = true;
+
+  // Should be true if a handshaker requires that Envoy sets the ALPN selection
+  // callback described by `SSL_CTX_set_alpn_select_cb`.
+  bool should_set_alpn_select_cb = true;
+
+  // Should be true if a handshaker requires that Envoy sets the TLS extension
+  // ticket key callback described by `SSL_CTX_set_tlsext_ticket_key_cb`.
+  bool should_set_tlsext_ticket_key_cb = true;
+
+  // Should be true if a handshaker requires that Envoy sets the SSL timeout as
+  // described by `SSL_CTX_set_timeout`.
+  bool should_set_timeout = true;
+
+  // Should be true if a handshaker requires that Envoy set the list of ciphers,
+  // as described by `SSL_CTX_set_strict_cipher_list`.
+  bool should_set_strict_cipher_list = true;
+
+  // Should be true if a handshaker requires that Envoy set the list of curves,
+  // as described by `SSL_CTX_set1_curves_list`.
+  bool should_set1_curves_list = true;
+
+  // Should be true if a handshaker requires that Envoy verify certificates.
+  bool should_verify_certificates = true;
+
+  // Should return true if this handshaker is FIPS-compliant.
+  // Envoy will fail to compile if this returns true and `--define=boringssl=fips`.
+  bool is_fips_compliant = true;
+};
```

Comment:
This is way too fine-grained, IMHO (it should be expressed as a list of custom handshaker capabilities, e.g. `provides_certificates`, `handles_session_resumption`, `handles_peer_verification`, rather than a list of BoringSSL functions to call, otherwise this will be extremely sensitive to any changes).

But looking at this now, I think we should move all those configuration settings to the handshaker, otherwise it has no way to learn about them (e.g. which cipher suites it should use, etc.).

Reply:
I don't think it makes sense to move these configurations to the handshaker, since no handshaker object exists during the instantiation of (for example) `ServerContextConfigImpl`, which needs to know which of these features it needs to fulfill and which of these features it can leave to the handshaker itself. The current design, which exposes these features as a part of the factory which creates the handshaker, solves that problem.

I agree that these permissions are too fine-grained and brittle. On that note, I've renamed this `HandshakerCapabilities` and reversed the polarity of the booleans; handshaker implementations now get to declare that they have the capability to handle feature Foo natively; otherwise Envoy will handle it as normal. The default implementation obviously has none of these capabilities, and defers to Envoy. 

I have also renamed and condensed these capabilities somewhat to make them more general. PTAL.